compiler/cpp/test/ReaderTest.cpp

```diff
@@ -0,0 +1,234 @@
+    reader.advance();
+    reader.advance();
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  't');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+}
+
+TEST(ReaderTests, EoFInEmptyFile) {
+    // Initialize
+    Reader reader = Reader("", 5);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  EOF);
+}
+
+// ----------------------------------------------------- Lookahead -----------------------------------------------------
+
+TEST(ReaderTests, Lookahead) {
+    // Initialize
+    Reader reader = Reader("File input", 2);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  'F');
+}
+
+TEST(ReaderTests, LookaheadAdvance) {
+    // Initialize
+    Reader reader = Reader("File input", 2);
+
+    // Test
+    EXPECT_EQ (reader.getLookahead(),  'F');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'i');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'l');
+    reader.advance();
+    EXPECT_EQ (reader.getLookahead(),  'e');
+}
+
+// --------------------------------------------------- Max lookahead ---------------------------------------------------
+
+TEST(ReaderTests, MaxLookahead) {
+    // Initialize
+    Reader reader = Reader("File input", 4);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "File");
+}
+
+TEST(ReaderTests, MexLookaheadAdvance) {
+    // Initialize
+    Reader reader = Reader("File input", 3);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "Fil");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "ile");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "le ");
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "e i");
+}
+
+TEST(ReaderTests, MaxLookaheadInvalid) {
+    // Initialize & test
+    try {
+        Reader reader = Reader("Test", 0);
+        FAIL() << "Expected MaxLookaheadException";
+    } catch(MaxLookaheadException const & err) {
+        EXPECT_EQ(err.what(),std::string("Invalid maximum lookahead '0' specified. Must be >= 1"));
+    }
+}
+
+TEST(ReaderTests, MaxLookaheadIsOne) {
+    // Initialize
+    Reader reader = Reader("File input", 1);
+
+    // Test
+    EXPECT_EQ (reader.getMaxLookahead(),  "F");
+    EXPECT_EQ (reader.getMaxLookahead(),  std::string(1, reader.getLookahead()));
+    reader.advance();
+    reader.advance();
+    reader.advance();
+    EXPECT_EQ (reader.getMaxLookahead(),  "e");
+    EXPECT_EQ (reader.getMaxLookahead(),  std::string(1, reader.getLookahead()));
+}
+
+// ------------------------------------------------------- Expect ------------------------------------------------------
+
+TEST(ReaderTests, ExpectSuccess) {
+    // Initialize
+    Reader reader = Reader("File input", 4);
+
+    // Test
+    try {
```

Comment:
For the tests: can't you just let the test method throw an error so that you don't have to catch it? Or is this not possible?

Reply:
As far as I know, you shouldn't do that

Reply:
@marcauberer  Why not in test files? I mean: this would even have the advantage that your error message gets printed. This is not the case, right now, e.g.:

https://github.com/compose-generator/ccom/blob/0db3276804dd898c79cfb86dc8633df181825df0/compiler/cpp/test/ReaderTest.cpp#L159-L177

which will not print your actual error message but rather the string you've defined in lin 175.

Reply:
Changed it